accord-core/src/main/java/accord/coordinate/AsynchronousAwait.java

```
@@ -0,0 +1,140 @@
+        this.synchronousCallback = synchronousCallback;
+    }
+
+    /**
+     * we require a Route to contact so we can be sure a home shard recipient invokes {@link Commands#updateRouteOrParticipants},
+     * notifying the progress log of a Route to determine it is the home shard.
+     */
+    public static AsynchronousAwait awaitAny(Node node, Topologies topologies, TxnId txnId, Route<?> contact, BlockedUntil awaiting, int asynchronousCallbackId, BiConsumer<SynchronousResult, Throwable> synchronousCallback)
+    {
+        AwaitTracker tracker = new AwaitTracker(topologies);
+        AsynchronousAwait result = new AsynchronousAwait(txnId, contact, tracker, asynchronousCallbackId, synchronousCallback);
+        result.start(node, topologies, contact, awaiting);
+        return result;
+    }
+
+    private void start(Node node, Topologies topologies, Route<?> route, BlockedUntil blockedUntil)
+    {
+        node.send(topologies.nodes(), to -> new Await(to, topologies, txnId, route, blockedUntil, asynchronousCallbackId), this);
+    }
+
+    @Override
+    public synchronized void onSuccess(Id from, AwaitOk reply)
+    {
+        if (isDone) return;
+
+        if (tracker.recordSuccess(from, reply == AwaitOk.Ready) == Success)
+            onSuccess();
+    }
+
+    @Override
+    public synchronized void onFailure(Id from, Throwable failure)
+    {
+        if (isDone) return;
+
+        this.failure = FailureAccumulator.append(this.failure, failure);
+        RequestStatus status = tracker.recordFailure(from);
+        switch (status)
+        {
+            default: throw new AssertionError("Unhandled RequestStatus: " + status);
+            case NoChange: break;
+            case Success:
+                onSuccess();
+                break;
+            case Failed:
+                isDone = true;
+                synchronousCallback.accept(null, this.failure);
+        }
+    }
+
+    private void onSuccess()
+    {
+        isDone = true;
```

I found this ordering of `isDone` first to be problematic because if we have an exception thrown in `synchronousCallback` or elsewhere then `onCallbackFailure` will suppress the failure because `isDone` is true. I changed it in one place in `ReadData`.

I think we should figure out how this is supposed to work and do a pass over all `isDone` handling to make it consistent.

---

That is intentional. You can't safely invoke the callback more than once. We could have a special case in all `onCallbackFailure` implementations where we invoke `unhandledException`.